src/ServiceControl/Notifications/Email/SendEmailNotificationHandler.cs

```diff
@@ -0,0 +1,115 @@
+﻿namespace ServiceControl.Notifications.Email
+{
+    using System;
+    using System.Threading.Tasks;
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using NServiceBus.Transport;
+    using Raven.Client;
+    using ServiceBus.Management.Infrastructure.Settings;
+
+    public class SendEmailNotificationHandler : IHandleMessages<SendEmailNotification>
+    {
+        readonly IDocumentStore store;
+        readonly EmailThrottlingState throttlingState;
+
+        public SendEmailNotificationHandler(IDocumentStore store, Settings settings, EmailThrottlingState throttlingState)
+        {
+            this.store = store;
+            this.throttlingState = throttlingState;
+
+            emailDropFolder = settings.EmailDropFolder;
+        }
+
+        public async Task Handle(SendEmailNotification message, IMessageHandlerContext context)
+        {
+            NotificationsSettings notifications;
+
+            using (var session = store.OpenAsyncSession())
+            {
+                notifications = await session.LoadAsync<NotificationsSettings>(NotificationsSettings.SingleDocumentId)
+                    .ConfigureAwait(false);
+            }
+
+            if (notifications == null || !notifications.Email.Enabled)
+            {
+                log.Info("Skipping email sending. Notifications turned-off.");
+                return;
+            }
+
+            var hasSemaphore = false;
+
+            try
+            {
+                while (hasSemaphore == false)
+                {
+                    if (throttlingState.IsThrottling())
+                    {
+                        log.Warn("Email notification throttled");
+                        return;
+                    }
+
+                    hasSemaphore = await throttlingState.Semaphore.WaitAsync(spinDelay).ConfigureAwait(false);
+                }
+
+                if (context.MessageId == throttlingState.RetriedMessageId)
+                {
+                    message.Body +=
+                        "\n\nWARNING: Your SMTP server was temporarily unavailable. Make sure to check ServicePulse for a full list of health check notifications.";
+                }
+
+                await EmailSender.Send(notifications.Email, message.Subject, message.Body, emailDropFolder)
+                    .ConfigureAwait(false);
+            }
+            catch (Exception e) when (!(e is OperationCanceledException))
+            {
+                if (message.IsFailure)
+                {
+                    throttlingState.ThrottlingOn();
+
+                    await Task.Delay(throttlingDelay).ConfigureAwait(false);
```

Comment:
Wouldn't this require a cancellation token for shutdown scenarios to make sure we don't hang in there? Also aren't we risking running into message timeout scenarios? For example, the default visibility timeout for SQS is 30 seconds.